packages/backend/src/open_payments/wallet_address/middleware.ts

```diff
@@ -10,20 +10,56 @@ export function createWalletAddressMiddleware() {
     ctx: AppContext,
     next: () => Promise<unknown>
   ): Promise<void> => {
-    if (
-      ctx.path === '/incoming-payments' ||
-      ctx.path === '/outgoing-payments' ||
-      ctx.path === '/quotes'
-    ) {
-      if (ctx.method === 'GET') {
-        ctx.walletAddressUrl = ctx.query['wallet-address'] as string
-      } else if (ctx.method === 'POST') {
-        ctx.walletAddressUrl = (ctx.request.body as CreateBody).walletAddress
+    if (ctx.method === 'GET') {
+      if (ctx.path && ctx.path.startsWith('/incoming-payments')) {
+        const incomingPaymentService = await ctx.container.use(
+          'incomingPaymentService'
+        )
+        const incomingPayment = await incomingPaymentService.get({
+          id: ctx.params.id
+        })
+        if (!incomingPayment || !incomingPayment.walletAddress) {
+          ctx.throw(401)
```

Comment:
401 as to not give away whether the resource exists or not?

Reply:
I wasn't really thinking that through but I think that counts as reasoning ;-)